lib/checkstl.cpp

```diff
@@ -1237,6 +1237,148 @@ void CheckStl::if_findError(const Token *tok, bool str)
+    if (Token::simpleMatch(top, "=") && Token::simpleMatch(top->astOperand1(), "[")) {
+        icontainerTok = top->astOperand1()->astOperand1();
+        ikeyTok = top->astOperand1()->astOperand2();
+        ivalueTok = top->astOperand2();
+    }
+    if (Token::simpleMatch(top, "(") && Token::Match(top->astOperand1(), ". insert|emplace (") && !astIsIterator(top->astOperand1()->tokAt(2))) {
+        icontainerTok = top->astOperand1()->astOperand1();
+        const Token *itok = top->astOperand1()->tokAt(2)->astOperand2();
+        if (Token::simpleMatch(itok, ",")) {
+            ikeyTok = itok->astOperand1();
+            ivalueTok = itok->astOperand2();
+        } else {
+            ikeyTok = itok;
+        }
+    }
+    if (!ikeyTok || !icontainerTok)
+        return nullptr;
+    if (isSameExpression(true, true, containerTok, icontainerTok, library, true, false) &&
+        isSameExpression(true, true, keyTok, ikeyTok, library, true, true)) {
+        if (ivalueTok)
+            return ivalueTok;
+        else
+            return ikeyTok;
+    }
+    return nullptr;
+}
+
+void CheckStl::checkFindInsert()
+{
+    if (!mSettings->isEnabled(Settings::PERFORMANCE))
+        return;
+
+    const SymbolDatabase *const symbolDatabase = mTokenizer->getSymbolDatabase();
+    for (const Scope *scope : symbolDatabase->functionScopes) {
+        for (const Token *tok = scope->bodyStart->next(); tok != scope->bodyEnd; tok = tok->next()) {
+            if (!Token::simpleMatch(tok, "if ("))
+                continue;
+            if (!Token::simpleMatch(tok->next()->link(), ") {"))
+                continue;
+            if (!Token::Match(tok->next()->astOperand2(), "%comp%"))
+                continue;
+            const Token *condTok = tok->next()->astOperand2();
+            const Token *containerTok;
+            const Token *keyTok;
+            std::tie(containerTok, keyTok) = isMapFind(condTok->astOperand1());
+            if (!containerTok)
+                continue;
+
+            const Token *thenTok = tok->next()->link()->next();
+            const Token *valueTok = findInsertValue(thenTok, containerTok, keyTok, mSettings->library);
+            if (!valueTok)
+                continue;
+
+            if (Token::simpleMatch(thenTok->link(), "} else {")) {
+                const Token *valueTok2 =
+                    findInsertValue(thenTok->link()->tokAt(2), containerTok, keyTok, mSettings->library);
+                if (!valueTok2)
+                    continue;
+                if (isSameExpression(true, true, valueTok, valueTok2, mSettings->library, true, true)) {
+                    checkFindInsertError(valueTok);
+                }
+            } else {
+                checkFindInsertError(valueTok);
+            }
+        }
+    }
+}
+
+void CheckStl::checkFindInsertError(const Token *tok)
+{
+    reportError(
+        tok, Severity::performance, "stlFindInsert", "Searching before insertion is not necessary.", CWE398, false);
```

Comment:
what do you think about something like:

    The code 'if (x.find(y)) x.insert(y);' is logically equivalent to 'x.insert(y);'

This would reuse the text somewhat from `checkDuplicateConditionalAssign`.

Reply:
I am not sure how easy that is. We check for other scenarios. For example, `if(x.find(y) != x.end()) x.insert(y)` is logically equivalent `x.insert(y)` but it is also logically equivalent to removing the code completely(and perhaps could be a bug in the code). Currently, I dont check if the search is checking if it is found or not found. Either way, the search is unnecessary.

Also, for scenarios like `if(x.find(y) == x.end()) x.emplace(y, 1); else x[y] = 1;` this is logically equivalent to `x[y] = 1`, so some deeper analysis is needed to distinguish these scenarios.